inc/xmath.h

```diff
@@ -1449,116 +1471,214 @@ struct quat : public vec<4, QS>
+	const auto& p_n = plane_n;
+	const auto& p_o = plane_o;
+	//
+	// definition of point on ray
+	// r_d * t + r_o = p
+	const auto& r_d = ray_d;
+	const auto& r_o = ray_o;
+
+	// substitute ray point into plane
+	// p_n . (p_o - (r_d * t + r_o)) = 0
+	// p_n . p_o - p_n * r_p = 0
+	// p_n . p_o = p_n * r_p         # expand r_p
+	// p_n . p_o = p_n . r_d * t + p_n . r_o
+	// (p_n . p_o) - (p_n . r_o) = p_n . r_d * t
+	// p_n . (p_o - r_o) / p_n . r_d = t
+
+	auto pn_dot_rd = p_n.dot(r_d);
+	auto d_po_ro   = p_o - r_o;
+
+	if (d_po_ro.dot(r_d) < 0)
+	{
+		// ray origin is on the side facing away from the plane normal
+		// an intersection cannot occur
+		return NAN;
+	}
+
+	if (pn_dot_rd >= 0)
+	{
+		// ray direction is parallel, or pointing away from the plane
+		// an intersection cannot occur
+		return NAN;
+	}
+
+	return p_n.dot(d_po_ro) / pn_dot_rd;
+}
+
+static XMTYPE ray_box(const vec<3>& ray_o,
+                      const vec<3>& ray_d,
+                      const vec<3>& box_o,
+                      const vec<3>  box_sides[3])
+{
+	const auto epsilon = std::numeric_limits<XMTYPE>::epsilon();
+	auto       t_min   = -INFINITY;
+	auto       t_max   = INFINITY;
+
+	auto p = box_o - ray_o;
+
+	XMTYPE half_lengths[] = {
+	    box_sides[0].magnitude(),
+	    box_sides[1].magnitude(),
+	    box_sides[2].magnitude(),
+	};
+
+	for (unsigned i = 0; i < 3; i++)
+	{
+		auto e = p.dot(box_sides[i] / half_lengths[i]);
+		auto f = ray_d.dot(box_sides[i] / half_lengths[i]);
+
+		if (static_cast<XMTYPE>(fabs(f)) > epsilon)
 		{
-			// ray origin is on the side facing away from the plane normal
-			// an intersection cannot occur
-			return NAN;
+			auto t_1 = (e + half_lengths[i]) / f;
+			auto t_2 = (e - half_lengths[i]) / f;
+
+			if (t_1 > t_2) { std::swap(t_1, t_2); }
+			if (t_1 > t_min) { t_min = t_1; }
+			if (t_2 < t_max) { t_max = t_2; }
+			if (t_min > t_max) { return NAN; }
+			if (t_max < 0) { return NAN; }
 		}
-
-		if (pn_dot_rd >= 0)
+		else if ((-e - half_lengths[i]) > 0 || (-e + half_lengths[i]) < 0)
 		{
-			// ray direction is parallel, or pointing away from the plane
-			// an intersection cannot occur
 			return NAN;
 		}
-
-		return p_n.dot(d_po_ro) / pn_dot_rd;
 	}
 
-	static XMTYPE ray_box(const vec<3>& ray_o,
-	                      const vec<3>& ray_d,
-	                      const vec<3>& box_o,
-	                      const vec<3>  box_sides[3])
-	{
-		const auto epsilon = std::numeric_limits<XMTYPE>::epsilon();
-		auto       t_min   = -INFINITY;
-		auto       t_max   = INFINITY;
+	if (t_min > 0) { return t_min; }
 
-		auto p = box_o - ray_o;
+	return t_max;
+}
 
-		XMTYPE half_lengths[] = {
-		    box_sides[0].magnitude(),
-		    box_sides[1].magnitude(),
-		    box_sides[2].magnitude(),
-		};
+static XMTYPE ray_sphere(const vec<3>& ray_o,
+                         const vec<3>& ray_d,
+                         const vec<3>& sphere_o,
+                         const XMTYPE  sphere_r)
+{
+	auto l  = sphere_o - ray_o;
```

Comment:
Is this from the WP article: https://en.wikipedia.org/wiki/Line%E2%80%93sphere_intersection ? It looks like it could be, but it would still be good to be sure so you can check it at a later date.